### ISSUE TITLE
MutableValueGraph add/remove a Collection of nodes

### ISSUE THREAD

**Issue by piegamesde** [google/guava#3034]
I just wanted to remove a lot of nodes from my `MutableValueGraph` and wrote the following code:

    Set<ContextState> nodes = graph.nodes();
    nodes.retainAll(nodeSet);

I got a `java.lang.UnsupportedOperationException` because `nodes` is immuable, as I found out after some digging in the source code. Not only does this behavior differ from what I am used to from other data structures like `map.keySet()`, but it also lacks any hint that it does so. My proposals are:

 - Don't make `nodes()` return an immutable Set, but instead a mutable set whose changes are reflected on the base graph.
 - Add some convenience methods for adding and removing multiple nodes
 - Add some documentation to warn that this does not behave as one might excpect

Same might also apply to edges and other graph types in certain ways, but I haven't checked them out yet.

**Comment by jrtom** [comment]
First, a clarification: if the graph is mutable, `nodes()` does not return an immutable set (i.e., one whose entries are guaranteed to never change), it returns an unmodifiable view.  That is, if the nodes of the graph change in the usual ways, the `Set` returned by `nodes()` will reflect those changes.  This is a pretty common behavior in the Guava implementations.

This behavior is documented here: https://github.com/google/guava/wiki/GraphsExplained#accessor-behavior
(GraphsExplained is itself linked from the class-level Javadoc of the graph interfaces.)
If you have suggestions for where else this documentation might be useful, we're open to suggestions.

The problem with letting `nodes()`'s return value be modifiable is that removing nodes from the graph has side effects on the rest of the graph (in particular, all edges incident to the removed nodes must also be removed).

If that's the behavior you want, that's what the [`Graphs.inducedSubgraph`](https://google.github.io/guava/releases/22.0/api/docs/com/google/common/graph/Graphs.html#inducedSubgraph-com.google.common.graph.Graph-java.lang.Iterable-) method is for. 
This will return a new graph whose node set is restricted to the set you input.

If you want to modify your existing graph in place, then removing all the specified nodes is trivial:

```java
for (N node : nodeSet) {
  graph.removeNode(node);
}
```
We could create a utility method for that but it wouldn't be much shorter.

If you just want to create a Set that intersects `graph.nodes()` with `nodeSet`, i.e., you don't want to modify the input graph at all, then `Sets.intersection()` is likely what you want.

**Comment by piegamesde** [comment]
Thank you for this information. The only thing I am still missing now (a bit) is some convenience methods for bulk adding and removing nodes and edges. But I can understand if you don't think they provide enough benefit to add them in (I am a really lazy person so I always prefer them to iterators / for loops).

**Comment by jrtom** [comment]
I can understand your desire for convenience methods.  The problem is that there's kind of a slippery slope.

Adding/removing a collection of nodes: that's easy, we could do that; the API would be straightforward:

```java
mutableGraph.addNodes(Iterable<N> nodes);
mutableGraph.removeNodes(Iterable<N> nodes);
```

I would still have some reservations about removing a collection of nodes because of the side effect (removing the incident edges), but one could argue that that's no more confusing than removeNode() already is in that respect.

Removing a collection of edges would be a bit more awkward:
```java
mutableGraph.removeEdges(Iterable<EndpointPair<N>> edges);  // Graph, ValueGraph
mutableGraph.removeEdges(Iterable<E> edges); // Network
```

mostly because in the case of `Graph` and `ValueGraph` that it's harder to construct the `EndpointPair` objects you need to specify the edges being removed.

Where it gets really sticky is when you want to add a collection of edges.  For `Graph`s, it's no worse, but for `ValueGraph`s or `Network`s you have to provide a way of specifying values/objects for each pair of endpoints to be connected.

```java
mutableGraph.addEdges(Iterable<EndpointPair<N>> edges);
mutableValueGraph.addEdges(Map<EndpointPair<N>, V> endpointsToValue); 
mutableNetwork.addEdges(Multimap<EndpointPair<N>, E> endpointsToEdge);
```

For both adding and removing a collection of edges--especially adding--you would really have to ask yourself why you spent the effort to create these structures rather than just modifying the graph directly.  There might be contexts in which you would be handed such a Map/Multimap/Iterable, but my guess is that they'd be very rare, indeed.

So while we could support adding/removing collections of nodes and edges, it's my evaluation that they wouldn't provide much value, and wouldn't be very easy to work with (on the edges side).  We could hypothetically _just_ support adding/removing collections of nodes, but the asymmetry would be odd, at least.

(I should consider putting this on the common.graph FAQ; you're not the first person to ask for this.  :) )

**Comment by piegamesde** [comment]
Thank you for the explanation.

**Comment by oliviercailloux** [comment]
I disagree with the conclusion here. I think `graph.nodes().remove(n)`, `graph.nodes.add(n)` and `graph.edges().remove(e)` all are very natural. (I agree that `graph.edges().add(e)` is unlikely to be used, for the reason you state, and I agree that if part of these functionalities is implemented, then the rest must be as well.)

I see no drawback and no slippery slope with the suggestion of implementing this. It does not clutter the API at all, as no methods have to be added, and the resulting API would only be cleaner as the `Set` of nodes and of edges will better match its contract. The only problem would be, obviously, the time it takes to implement it, and I understand that you do not want to treat it as a high-priority feature. But it would be nice to have and I see no objection in principle.

In the meantime, please consider documenting the intended behavior on the `Graph#nodes()` method (possibly saying explicitly that it is not guaranteed that mutation of the resulting set is either possible or will be rejected, to leave options open for the future). I also didn’t find the doc.

**Comment by jrtom** [comment]
There are two separate issues here which have been kind of lumped into a single issue:

1.    should we provide a way of bulk-adding or bulk-removing nodes and edges
1.    should the `Set`s returned by `nodes()` and `edges()` be modifiable views instead of unmodifiable

I think we've explained why we don't currently feel that bulk-adding or bulk-removing nodes or edges is an API we want to support at this time (and see also my comment about how if you want to bulk-remove nodes that `Graphs.inducedSubgraph()` does that once you've defined the node set you want).

Note that if we made `nodes()` and `edges()` into modifiable views, then we would (by default) also have bulk-adding and bulk-removing methods.

As to why we don't want to go with modifiable views:

> the resulting API would only be cleaner as the `Set`...will better match its contract

The `Set` Javadoc specifically states that the mutation operations are optional: https://docs.oracle.com/javase/8/docs/api/java/util/Set.html#add-E-
that is, one should never assume that a given `Set` supports mutation.
(This is why the various graph interfaces do not include mutation methods, because I wanted to remove that ambiguity.)

The reason why `Graph.nodes()` et al. do not document that they return unmodifiable views is specifically because it is perfectly legal for someone to create a `Graph` implementation that returns modifiable views.  We have simply chosen, in our implementations, not to do so (for reasons outlined in https://github.com/google/guava/wiki/GraphsExplained#accessor-behavior).

The fact that our implementation classes are not public may contribute to confusion on this point, certainly.  If you have other suggestions for how to make this documentation more discoverable, I'd like to hear them.

**Comment by oliviercailloux** [comment]
My point is that the point 1. (bulk-adding and removing) should be made more precise, to avoid confusion. There’s 1a. bulk-adding and bulk-removing through the existing `nodes()` and `edges()` methods, and 1b. adding methods like `mutableGraph.addNodes(Iterable<N> nodes);`, `mutableGraph.removeNodes(Iterable<N> nodes);`, and so on, as you explained previously.

Now, 1a is equivalent to 2.

As I understood it, you previously argued against 1b, and I agree that 1b would be a bad option. I argue in favor of 1a (or equivalently, 2). I don’t think that you have explained why 1a / 2 is a bad idea, or at least, I thought I was replying to your arguments in my previous post.

I agree, the mutation operations are optional, and my phrasing was sloppy. I meant that it’s nicer when a structure also supports optional operations. I agree that it’s someone’s right, per the contract, to not provide these operations, my point is simply that I see no drawback and no slippery slope with the suggestion of implementing 1a / 2. But I see an advantage: implementing it would provide `graph.nodes().remove(n)`, `graph.nodes.add(n)`, `graph.edges().remove(e)`, `graph.edges().add(e)`, `graph.nodes().removeAll`, `graph.nodes.addAll`, `graph.edges().removeAll`, and so on, some of which at least I think you’ll agree feel very natural to use.

The statement [here](https://github.com/google/guava/wiki/GraphsExplained#accessor-behavior) about modifiable view being “a hazardous design choice” is precisely the point I disagree with in my post and I try to argue about. (To be precise: if the point is simply that extreme caution is required when implementing it, then I agree but it’s not an argument against implementing it, I’d say: I’m sure people at Google know how to code it carefully enough.) Modifiable views are very frequent in Java collections, including the ones added by Guava, and are very pleasant to use.

**Comment by jrtom** [comment]
Thanks for taking the time to respond; I think that we're zeroing in on some good points here.

I agree that it would be technically possible to implement a modifiable view of `nodes()` and `edges()`[1], although I think I'm less sold on the benefits, especially of `removeAll()` and `retainAll()` for edges in `Graph` and `ValueGraph`.  

However, I think that there are two more good reasons why this is not a good idea.

First, I think that some people would get confused about the mutability of those `Set`s.   That is, they might expect that they were mutable _copies_ rather than _views_, and that their modifications would not have any effect on the original graph.  This would be a bad mistake to make.

Second, and most importantly: `Graph` and its siblings are intentionally intrinsically not modifiable; modification is left to the subinterfaces `MutableGraph`, etc.

By making `nodes()` and `edges()` modifiable, we'd be letting in mutations by the side door, and I don't think that we want to go that direction with the design:  

*    If you know that your graph is a `MutableGraph`, it's not better (or shorter) to call `graph.nodes().add(x)` rather than `graph.addNode(x)`.  
*    If you _don't_ know that it's a `MutableGraph`, you don't want to find that out by trying to modify it; you'd create a mutable copy and operate on that.  (Or try casting it to a `MutableGraph` first, and creating a copy if that doesn't work.)

[1] Caveat: if that were to apply to the `Sets` returned by the accessor methods--that is, if they were to be made modifiable views as well--that would require us to abandon use of `Sets.union()` and `Sets.intersection()`, which return unmodifiable views (and are used in `adjacentNodes()`,  `incidentEdges()`, `edgesConnecting()`, etc.).  I haven't thought all the way through the technical hurdles involved in reimplementing that functionality to support modification, but it might be feasible.

**Comment by oliviercailloux** [comment]
Yes, interesting discussion!

You are absolutely right, there is a distinction to be made on `Graph` VS `MutableGraph` which I overlooked.

I’ll talk about `nodes()` only, the reasoning is the same for `edges()`.

In the solution of my dreams, `Graph#nodes()` would state that there is no guarantee that the returned set is mutable, and no guarantee that it is immutable. `MutableGraph#nodes()` would state that the returned set is a modifiable view. `ImmutableGraph#nodes()` states that the returned set is immutable.

I disagree that users would be surprised. It is very natural that `MutableGraph#nodes()` would return a modifiable view, because this is the usual way such methods behave. Examples are very numerous in [Java](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.html#keySet()) and [Guava](https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Table.html#rowKeySet--) collections, and I believe it would be hard to find counter-examples in well-known collection classes. It is also very natural that when some class has mutable and immutable variants, some of its methods change behavior from providing mutable views to providing immutable structures: again, this is most usual in existing classes. Finally, it is very natural that the `Graph` variant, which guarantees neither mutability nor immutability, would have its methods return structures with similar lack of guarantees.

I agree that the benefits are not immense, but they still exist. Also, when views are mutable, more patterns are available. For example, you can pass a set of nodes to another method, instead of passing the whole graph, if it only deals with adding and removing nodes, thereby limiting coupling and increasing clarity. (I admit I don’t have a concrete example in mind where it would be clearly useful.) The OP at least provides a demand that would be satisfied by modifiable views, and I recall having hoped for it while programming with graphs some time ago, though do not remember the specifics.

I think the alternative proposal is worst. If I understood correctly, the alternative proposal is the following. Both `Graph` and `MutableGraph` would provide no guarantee (of either mutability or immutability) about the returned set of nodes, while (of course) `ImmutableGraph` returns immutable sets. The argument is that some specific implementation (not Guava’s) is free (and welcome) to offer such mutability. But why not provide it as a guarantee in the `MutableGraph` interface itself, to provide a standard, if we agree it has no drawback? A contract with more guarantees is always better, IMHO, when it implies no drawback on other aspects of the structure.

To summarise my position, I understand, as I wrote, that you would not treat this demand with high priority (that’s obviously for you to decide anyway), but I think it would be nice to have and I see no objections to it on the aspect of quality of API, quite the contrary.

It is my hope that Guava will provide a very high quality set of Graph interfaces, because it is in a good position (thanks to its recognized general quality of API) to set a de-facto standard on graph structures that would create a favorable environment for easily interchangeable libraries working over graphs.

**Comment by Maaartinus** [comment]
@jrtom 
> If you want to modify your existing graph in place, then removing all the specified nodes is trivial:

Yes, but `retainAll` is a bit tricky. It could naively be done with
```
for (N node : Sets.difference(graph.nodes(), nodeSet)) {
  graph.removeNode(node);
}
```
but this calls for a `ConcurrentModificationException` (is `nodes()` a snapshot or a life view?). At the moment I can't see any simple solution (apart from e.g., `ImmutableList.copyOf`). A possibility to do `graph().nodes().retainAll(nodeSet)` would make this trivial.

---

@oliviercailloux 
> A contract with more guarantees is always better

... except for the implementors. But, I guess, given that Guava itself provides a good implementation of the `MutableGraph`, the number of other implementors will be pretty close to zero.

Adding any contract to an interface is actually a breaking change (source and binary compatible, but making old implementations break the new contract), so it should be *announced* ASAP (even when the implementation gets low priority).

> It is my hope that Guava will ... set a de-facto standard on graph structures

+1
